Feature ble host privacy

pan- opened the pull request:
### Summary of changes <!-- Required -->

This PR reimplements Bluetooth privacy for all type of Bluetooth controllers (4.0 to 5.2). 
As soon as privacy is enabled, a private address is used to advertise, scan or initiate connection and peers resolvable addresses are resolved if a bond with matching IRK is found in the security database.

This should solve most of the pairing issues observed on Android and iOS where the local device is not able to retrieve a previous bond.  

Internally there is two implementations of the address resolution: 
- Host based where the software compute if the address is known or not
- Controller based where the Bluetooth controller resolves automatically addresses received during advertising or scan. 

New configurations flags have been added to select at compile time the implementation to use: 
- `ble-gap-host-based-private-address-resolution` Forces the system to use host based resolution 
- `ble-gap-host-privacy-resolved-cache-size` Indicate how many entries must be booked for the host based resolution cache. 

This PR is the result of the following PRs merged on the feature branch
- Privacy feature configuration: https://github.com/ARMmbed/mbed-os/pull/13662
- Force device mode for LL resolution: https://github.com/ARMmbed/mbed-os/pull/13689
- Privacy enable event emission : https://github.com/ARMmbed/mbed-os/pull/13680
- Improvement of Secure DB entry retrieval: https://github.com/ARMmbed/mbed-os/pull/13660 and https://github.com/ARMmbed/mbed-os/pull/13658
- Force identity address to be static: https://github.com/ARMmbed/mbed-os/pull/13622
- Local private address management: https://github.com/ARMmbed/mbed-os/pull/13582 and https://github.com/ARMmbed/mbed-os/pull/13560
- Peer private address management: https://github.com/ARMmbed/mbed-os/pull/13630 and https://github.com/ARMmbed/mbed-os/pull/13685: 
- Fixes: https://github.com/ARMmbed/mbed-os/pull/13656 and https://github.com/ARMmbed/mbed-os/pull/13670
- Notification when advertising start or stop: https://github.com/ARMmbed/mbed-os/pull/13706 
  

#### Impact of changes <!-- Optional -->

Existing applications should not have anything to change regards to calls to privacy APIs as this PR enables it for all type of controllers and not just resolution when supported by the hardware. 

The advertising state was showing up as enabled as soon as the function `startAdvertising` was called. This is was not correct as it wasn't reflecting the reality of operations. The state is now modified when advertising is actually active. Two new events have been added to `Gap::EventHandler` to catch the change of advertising state.   

#### Migration actions required <!-- Optional -->

<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

None
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

[privacy-tests.zip](https://github.com/ARMmbed/mbed-os/files/5327764/privacy-tests.zip)

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

pan- committed:
BLE: Update ble::Duration to support default values

pan- committed:
ble: Add privacy mode to the pal

pan- committed:
BLE: Add resolvable_address_timeout_t datatype.

pan- committed:
BLE: Introduce PalPrivateAddressController

This PAL interface is responsible for generating, resolving private address and maintaining the controller or host resolving list.
It also indicates to upper level if LL resolution is supported or not and allows them to enable it.

pan- committed:
BLE: Add cordio implementation of PalPrivateAddressController

pan- committed:
BLE: Add PrivateAddressController module.

It controls private address generation and host or controller address resolution.

pan- committed:
BLE: Update security manager to use the PrivateAddressController

Resolution list has been removed from the pal as this is handled by the PrivateAddressController.

pan- committed:
BLE: Pal event queue pop before execution.

A race condition was present if a single event was present in the event queue and the event was generating a new event.

pan- committed:
BLE: Add gap handlers to be informed when advertising as started or been stopped

pan- committed:
BLE: Bind address controller with Gap

pan- committed:
BLE: Remove address resolution from GAP pal

pan- committed:
BLE: Add initiating flag into gap to track if the local device tries to connect to a peer.

pan- committed:
BLE: Add resolvable and non resolvable private address rotation with advertising

pan- committed:
BLE: rename address resolution handler.

pan- committed:
BLE: Add callback for scan started and stoped.

pan- committed:
BLE: Update state when scan started/stopped callback received.

pan- committed:
BLE: Fix random static address storage

pan- committed:
BLE: check LL resolution availability before enabling it.

pan- committed:
BLE: Add routine that computes the random address to use depending on the procedure.

pan- committed:
BLE: Implement private address rotation for initiating.

pan- committed:
BLE: Private address when initiating connection

pan- committed:
BLE: privacy tracing cleanup

pan- committed:
BLE: return stack busy instead of invalid state when applicable

enabling/disabling scanning and advertising operations can be prevented if the previous operations hasn't completed.
Return that the stack is busy and cannot fulfill the request at the moment rather than the state is invalid.

pan- committed:
BLE: Fix state update when event handler not present

pan- committed:
BLE: improve scan state bookkeeping

pan- committed:
Cordio: Add API to set the local identity address.

paul-szczepanek-arm committed:
split resolve addres into cache and not cache

pan- committed:
BLE: Add function to get the device random static address.

pan- committed:
BLE: Add function to set the identity address used by the controller/platform.

The function to get it has been removed as this operation is driven by the security manager.

pan- committed:
BLE: Set identity address to random static.

This also ensure the random static address used by gap is the correct one.

paul-szczepanek-arm committed:
event handler should not affect internal state

paul-szczepanek-arm committed:
remove unused enhanced conn complete

paul-szczepanek-arm committed:
Add setters to events

paul-szczepanek-arm committed:
add EventList class for storing pending events

paul-szczepanek-arm committed:
Add host reolustion for advertising and connections

pan- committed:
BLE: Check if extended advertising is available to clear advertising set upon connection

pan- committed:
BLE: Queue address resolution in event queue to avoid synchronous execution

Otherwise when no IRK are available, the process will execute synchronously, calling the callback before the return of the function.

pan- committed:
BLE: Check if the resolving list item is populated instead of IRK validity

pan- committed:
BLE: Fix insertion to host resolving list.

pan- committed:
BLE: Remove set advertising timeout from SM pal

pan- committed:
BLE: Update identity address when retrieved from the DB

pan- committed:
BLE: Update DB entry if current entry doesn't match requested EDIV and RAND

pan- committed:
BLE: Report connection internally when address resolution has completed.

paul-szczepanek-arm committed:
host privacy config option to enable it

pan- committed:
BLE: Add workaround for cordio use of global random address

In Cordio, the global random address is used when the local device is connected instead of the advertising set one.
This can cause pairing to fail if the advertising set address and the global random address are different.
This workaround forces the global random address to be set when advertising is started and extended advertising is present.

pan- committed:
BLE: Factorize peripheral privacy applied when connected

paul-szczepanek-arm committed:
add nrf override for host resolution

paul-szczepanek-arm committed:
filter based on resolving list size

paul-szczepanek-arm committed:
apply peripheral policy on ll privacy

paul-szczepanek-arm committed:
do not reject conn if no bond

paul-szczepanek-arm committed:
only apply policy if privacy enabled

pan- committed:
BLE: Add function to signal privacy initialization.

paul-szczepanek-arm committed:
allow host resolved address to be used for connection

paul-szczepanek-arm committed:
scan forever and allow cancellation

pan- committed:
Set privacy mode to device mode in LL.
Do not enable address resolution when a new entry is added.

paul-szczepanek-arm committed:
use enum for state

paul-szczepanek-arm committed:
use _initiating to block scan manipulation

pan- committed:
BLE: Notify application when advertising start or stop

Applicable for legacy and extended advertising.

mprse commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -487,6 +497,12 @@ ble_error_t Gap::connect(
+    if (_connect_to_host_resolved_address_state == ConnectionToHostResolvedAddressState::scan) {
+        return BLE_ERROR_OPERATION_NOT_PERMITTED;
+    }
+#endif BLE_GAP_HOST_BASED_PRIVATE_ADDRESS_RESOLUTION
```
`#endif //BLE_GAP_HOST_BASED_PRIVATE_ADDRESS_RESOLUTION`

pan- committed:
BLE: Workaround for WB55 connection event

The connection event reported by the WB55 is incorrect if controller privacy is not enable and the peer connects with an unknown private resolvable address: The RPA field contains the connection address (it should be empty) and the peer address is all FF while it should be equal to the connection address.

pan- committed:
BLE: Take advantage of controller based address resolution on WB55

paul-szczepanek-arm committed:
Fix conditional compilation based on config

paul-szczepanek-arm committed:
rename private address controller to match security manager

pan- committed:
Merge pull request #13759 from pan-/privacy-workaround-wb55

Privacy workaround wb55

paul-szczepanek-arm committed:
fix removed friend function

pan- committed:
Merge pull request #13760 from paul-szczepanek-arm/privacy-fix-ifdef

Fix conditional compilation for ble
